### PR TITLE
resolves #3665

### LINKED ISSUE #3665: Skip Attribute
Attributes are evolving and now are used also to create DSL. Now, by the rising of procedural macro and the possibility to use arbitrary rust code in attributes syntax formatting attribute by simple rules based on meta and nested meta can produce some formatting that make the procedural macro DSL not clear.

For instance I have a crate of mine [la10736/rstest](https://github.com/la10736/rstest) that you can use to write tests like this:

```rust
#[rstest_parametrize(foo, bar, baz,
    case(0, "a", "A"),
    case(1, "b", "B")
)]
fn mytest(foo: u32, bar: &str, baz: &str) {}
```
Now, use `rustfmt` on this file make it flat and hard to read:

```rust
#[rstest_parametrize(foo, bar, baz, case(0, 'a', 'A'), case(1, 'b', 'B'))]
fn mytest(foo: u32, bar: char, baz: char) {}
```

Or wrost:
```rust
#[rstest_parametrize(foo, bar, baz,
    case(0, "a", "Aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
    case(1, "b", "B")
)]
fn mytest(foo: u32, bar: &str, baz: &str) {}
```
in

```rust
#[rstest_parametrize(
    foo, 
    bar,
    baz,
    case(0, "a", "Aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
    case(1, "b", "B")
)]
fn mytest(foo: u32, bar: &str, baz: &str) {}
```

We can use `rustfmt::skip::macros()` or a brand new `rustfmt::skip::attributes()` to have the ability of skip attributes formatting too.

If you are ok I can work on it.

### PR DESCRIPTION
**Pull request by la10736**
resolves #3665

* Implementation and test based on `rustfmt::skip::macros()` feature
* Refactoring to encapsulate both macros and attributes
* Removed useless `RefCell`

### CODE REVIEW

**Comment by scampi** [comment]
can you update the changelog file ?

File: src/attr.rs [hunk 1]
```diff
@@ -319,9 +319,13 @@ impl Rewrite for ast::Attribute {
         if self.is_sugared_doc {
             rewrite_doc_comment(snippet, shape.comment(context.config), context.config)
         } else {
+            let should_skip = self
+                .ident()
+                .map(|s| context.skip_context.attributes_skip(&*s.name.as_str()))
```
[the remark this answers is not in the document]
Comment (la10736) [reply src/attr.rs]: I was wrong... `&s.name.as_str()` is enougth

**Comment by la10736** [comment]
Sorry... I did some craps on rebasing my fork.... Now should be ok.

**Comment by scampi** [comment]
@la10736 there are few more commits you can drop, e.g.,  `fix print-config minimal option`. It's ok to just open a new clean PR ;o)

**Comment by la10736** [comment]
Ok I'm closing it

**Comment by la10736** [comment]
Ok a fresh new PR at #3689

**Pull request closed without merging**